File: app/src/main/java/github/umer0586/fragments/ConnectionsFragment.java
package github.umer0586.fragments;
import android.content.ComponentName;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.java_websocket.WebSocket;

import java.util.ArrayList;

import github.umer0586.R;
import github.umer0586.fragments.customadapters.ConnectionsRecyclerViewAdapter;
import github.umer0586.websocketserver.ConnectionsChangeListener;
import github.umer0586.service.SensorService;
import github.umer0586.service.ServiceBindHelper;
import github.umer0586.util.UIUtil;

/**
 * TODO: functionality to allow user to close all connections (using button in action bar)
 */
public class ConnectionsFragment extends Fragment
        implements ServiceConnection, ConnectionsChangeListener {

    private static final String TAG = github.umer0586.fragments.ConnectionsFragment.class.getSimpleName();

    private SensorService sensorService;
    private ServiceBindHelper serviceBindHelper;

    private RecyclerView recyclerView;
    private ConnectionsRecyclerViewAdapter connectionsRecyclerViewAdapter;
    private ArrayList<WebSocket> webSockets = new ArrayList<>();

    private TextView noConnectionsText;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
        Log.d(TAG, "onCreateView()");
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_connections, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState)
    {
        super.onViewCreated(view, savedInstanceState);
        recyclerView = view.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        connectionsRecyclerViewAdapter = new ConnectionsRecyclerViewAdapter(webSockets);
        recyclerView.setAdapter(connectionsRecyclerViewAdapter);

        noConnectionsText = view.findViewById(R.id.no_connections_text);

        serviceBindHelper = new ServiceBindHelper(
                getContext(),
                this,
                SensorService.class
        );

        getLifecycle().addObserver(serviceBindHelper);
    }


    @Override
    public void onPause()
    {
        super.onPause();
        Log.d(TAG, "onPause()");

        if(sensorService != null)
            sensorService.setConnectionsChangeListener(null);

    }

    @Override
    public void onDestroy()
    {
        super.onDestroy();
        getLifecycle().removeObserver(serviceBindHelper);
    }


    @Override
    public void onServiceConnected(ComponentName name, IBinder service)
    {

        SensorService.LocalBinder localBinder = (SensorService.LocalBinder) service;
        sensorService =  localBinder.getService();

        if(sensorService != null)
        {
            sensorService.setConnectionsChangeListener(this);
            ArrayList<WebSocket> webSockets = sensorService.getConnectedClients();

            if(webSockets != null)
            {
                this.webSockets.clear();
                this.webSockets.addAll(webSockets);
            }
            handleNoConnectionsText();

            if(webSockets != null)
                connectionsRecyclerViewAdapter.notifyDataSetChanged();

        }


    }

    @Override
    public void onServiceDisconnected(ComponentName name)
    {

    }


    @Override
    public void onConnectionsChanged(ArrayList<WebSocket> webSockets)
    {

        this.webSockets.clear();
        this.webSockets.addAll(webSockets);
        handleNoConnectionsText();

        UIUtil.runOnUiThread(()->{
            connectionsRecyclerViewAdapter.notifyDataSetChanged();
        });
    }

    private void handleNoConnectionsText()
    {
        if(webSockets.size() > 0)
            UIUtil.runOnUiThread(()-> noConnectionsText.setVisibility(View.INVISIBLE));
        else
            UIUtil.runOnUiThread(()-> noConnectionsText.setVisibility(View.VISIBLE));
    }


}

File: app/src/main/java/github/umer0586/sensorserver/fragments/ConnectionsFragment.kt
package github.umer0586.sensorserver.fragments

import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import github.umer0586.sensorserver.databinding.FragmentConnectionsBinding
import github.umer0586.sensorserver.fragments.customadapters.ConnectionsRecyclerViewAdapter
import github.umer0586.sensorserver.service.WebsocketService
import github.umer0586.sensorserver.service.WebsocketService.LocalBinder
import github.umer0586.sensorserver.service.ServiceBindHelper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.java_websocket.WebSocket

/**
 * TODO: functionality to allow user to close all connections (using button in action bar)
 */
class ConnectionsFragment : Fragment()
{

    private var websocketService: WebsocketService? = null
    private lateinit var serviceBindHelper: ServiceBindHelper

    private lateinit var connectionsRecyclerViewAdapter: ConnectionsRecyclerViewAdapter
    private val webSockets = ArrayList<WebSocket>()


    private var _binding : FragmentConnectionsBinding? = null
    private val binding get() = _binding!!

    override fun onCreateView( inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle? ): View?
    {
        Log.d(TAG, "onCreateView()")

        _binding = FragmentConnectionsBinding.inflate(inflater, container, false)
        val view = binding.root

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?)
    {
        super.onViewCreated(view, savedInstanceState)

        binding.recyclerView.setLayoutManager(LinearLayoutManager(context))

        connectionsRecyclerViewAdapter = ConnectionsRecyclerViewAdapter(webSockets)
        binding.recyclerView.setAdapter(connectionsRecyclerViewAdapter)


        serviceBindHelper = ServiceBindHelper(
            context = requireContext(),
            service = WebsocketService::class.java,
            componentLifecycle = lifecycle
        )

        serviceBindHelper.onServiceConnected(this::onServiceConnected)

    }

    override fun onPause()
    {
        super.onPause()
        Log.d(TAG, "onPause()")

        // To prevent memory leak
        websocketService?.onConnectionsChange( callBack = null)
    }


    fun onServiceConnected(binder: IBinder)
    {
        val localBinder = binder as LocalBinder
        websocketService = localBinder.service


        websocketService?.onConnectionsChange{ webSockets ->

            this.webSockets.clear()
            this.webSockets.addAll(webSockets)

            handleNoConnectionsText()
            lifecycleScope.launch(Dispatchers.Main) {
                connectionsRecyclerViewAdapter.notifyDataSetChanged()
            }
        }



        websocketService?.getConnectedClients().let { webSockets ->

            this.webSockets.clear()
            webSockets?.let{this.webSockets.addAll(it)}

            connectionsRecyclerViewAdapter.notifyDataSetChanged()

        }

        handleNoConnectionsText()



    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }


    private fun handleNoConnectionsText()
    {
        lifecycleScope.launch(Dispatchers.Main) {
            if (webSockets.size > 0)
                binding.noConnectionsText.visibility = View.INVISIBLE
            else
                binding.noConnectionsText.visibility = View.VISIBLE
        }
    }


    companion object
    {
        private val TAG: String = ConnectionsFragment::class.java.getSimpleName()
    }
}